datafusion/physical-expr/src/equivalence/projection.rs

```diff
@@ -156,6 +157,97 @@ impl FromIterator<(Arc<dyn PhysicalExpr>, ProjectionTargets)> for ProjectionMapp
+        projected_orderings.extend(project_ordering(ordering, schema));
+    }
+
+    projected_orderings
+}
+
+/// Projects a single [LexOrdering] onto the given schema.
+///
+/// This function attempts to rewrite every [PhysicalSortExpr] in the provided
+/// [LexOrdering] so that any [Column] expressions point at the correct field
+/// indices in `schema`.
+///
+/// Key details:
+/// - Columns are matched by name, not by index. The index of each matched
+///   column is looked up with [Schema::column_with_name](arrow::datatypes::Schema::column_with_name) and a new
+///   [Column] with the correct [index](Column::index) is substituted.
+/// - If an expression references a column name that does not exist in
+///   `schema`, projection of the current ordering stops and only the already
+///   rewritten prefix is kept. This models the fact that a lexicographical
+///   ordering remains valid for any leading prefix whose expressions are
+///   present in the projected schema.
+/// - If no expressions can be projected (i.e. the first one is missing), the
+///   function returns `None`.
+///
+/// Return value:
+/// - `Some(LexOrdering)` if at least one sort expression could be projected.
+///   The returned ordering may be a strict prefix of the input ordering.
+/// - `None` if no part of the ordering can be projected onto `schema`.
+///
+/// Example
+///
+/// Suppose we have an input ordering `[col("a@0"), col("b@1")]` but the projected
+/// schema only contains b and not a. The result will be `Some([col("a@0")])`. In other
+/// words, the column reference is reindexed to match the projected schema.
+/// If neither a nor b is present, the result will be None.
+pub fn project_ordering(
+    ordering: &LexOrdering,
+    schema: &SchemaRef,
+) -> Option<LexOrdering> {
+    let mut projected_exprs = vec![];
+    for PhysicalSortExpr { expr, options } in ordering.iter() {
+        let transformed = Arc::clone(expr).transform_up(|expr| {
+            let Some(col) = expr.as_any().downcast_ref::<Column>() else {
+                return Ok(Transformed::no(expr));
+            };
+
+            let name = col.name();
+            if let Some((idx, _)) = schema.column_with_name(name) {
+                // Compute the new column expression (with correct index) after projection:
+                Ok(Transformed::yes(Arc::new(Column::new(name, idx))))
+            } else {
+                // Cannot find expression in the projected_schema,
+                // signal this using an Err result
+                plan_err!("")
+            }
+        });
+
+        match transformed {
+            Ok(transformed) => {
+                projected_exprs.push(PhysicalSortExpr::new(transformed.data, *options));
+            }
+            Err(_) => {
+                // Err result indicates an expression could not be found in the
+                // projected_schema, stop iterating since rest of the orderings are violated
+                break;
```

Comment:
Should we give a `warn!()` here?

Reply:
It's not really an error case. This happens, for instance, when the input has sort order `a, b`, but the projection of the node only retains `a`. The resulting order is just `a`.

I'm kind of abusing `Err` here. The `plan_err!("")` bit is basically just there to abort the `transform_up` tree walk early since there's no point in continuing. The `Err(_)` bit is handling it. Not pretty, but I couldn't find a nicer way to achieve the desired result.

FWIW, this is the same behaviour as the code that's already present.